Objects/abstract.c

```diff
@@ -2625,6 +2625,9 @@ object_recursive_isinstance(PyThreadState *tstate, PyObject *inst, PyObject *cls
         return object_isinstance(inst, cls);
     }
 
+    if (_PyUnion_Check(cls))
+        cls = _Py_union_args(cls);
```

Comment:
```suggestion
    if (_PyUnion_Check(cls)) {
        cls = _Py_union_args(cls);
    }
```
https://peps.python.org/pep-0007/#code-lay-out